### PR TITLE
Deploy Jenkins on Elastic Beanstalk using CodePipeline and CodeBuild

### PR DESCRIPTION
**Pull request by aknysh**
## What

* `Terraform` module to build `Docker` image with `Jenkins`, save it to `ECR`, and deploy to `Elastic Beanstalk` running `Docker` stack 

* `terraform-aws-jenkins` module will:

   * Create `Elastic Beanstalk Application`
   * Create `Elastic Beanstalk Environment` running `Docker` stack
   * Create `ECR` `Docker` repository
   * Create `CodePipeline` with `CodeBuild`
   * After creating all the `AWS` resources, the `CodePipeline` will 
      * get the code from the specified  `Jenkins` repo
      * build `Docker` image from it
      * save the `Docker` image to the specified `ECR` repo
      * deploy the `Docker` image to `EB` running `Docker` stack

## Why

* Simple, highly available deployment of `Jenkins` so that even `Jenkins` itself follows the `CI/CD` pattern

### CODE REVIEW

File: variables.tf [hunk 1]
```diff
@@ -77,7 +77,7 @@ variable "autoscale_min" {
 }
 
 variable "autoscale_max" {
-  default = "1"
+  default = "2"
```
Comment (osterman) [inline variables.tf]: Is this safe? My concern would be to have more than one jenkins master online sharing the same FS at any given time.

Comment (aknysh) [reply variables.tf]: Was my concern as well.

`eb_environment` did not work when `autoscale_max` was set to `1` (Exception thrown by `AWS`).

> MinInstancesInService: Invalid option value: '1.0' (Namespace: 'aws:autoscaling:updatepolicy:rollingupdate', OptionName: 'MinInstancesInService'): The minimum instances in service for rolling updates (1) should be less than the maximum instance count for the Autoscaling group (1). 

Looks like we need to check the two values (`autoscale_min` and `autoscale_max`) and if `autoscale_min` >= `autoscale_max`, disable rolling updates.

Comment (aknysh) [reply variables.tf]: Actually, we don't need to change anything in `eb_environment` module.
Setting `updating_min_in_service` to `0` and `autoscale_max` to `1` worked.

File: variables.tf [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+  type    = "string"
+  default = "/"
+}
+
+variable "loadbalancer_type" {
+  type    = "string"
+  default = "application"
+}
+
+variable "eb_instance_type" {
```
Comment (osterman) [inline variables.tf]: rename to `master_instance_type`

**Comment by osterman** [comment]
add `description` to all attributes as TF registry will parse this and use it.

File: main.tf [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+  attributes = ["ecr"]
+  namespace  = "${var.namespace}"
+  name       = "${var.name}"
+  stage      = "${var.stage}"
+}
+
+# EFS to store Jenkins state (settings, jobs, etc.)
+module "efs" {
+  source             = "git::https://github.com/cloudposse/tf_efs.git?ref=tags/0.2.0"
+  attributes         = "efs"
```
Comment (osterman) [inline main.tf]: this should be a `list`

File: main.tf [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+module "elastic-beanstalk-application" {
+  source      = "git::https://github.com/cloudposse/terraform-aws-elastic-beanstalk-application.git?ref=tags/0.1.2"
+  namespace   = "${var.namespace}"
+  name        = "${var.name}"
+  stage       = "${var.stage}"
+  description = "${var.description}"
+}
+
+# Elastic Beanstalk Environment
+module "elastic-beanstalk-environment" {
```
Comment (osterman) [inline main.tf]: Use `_` instead of `-`

File: README.md [hunk 5]
```diff
@@ -1,2 +1,8 @@
-# tf_jenkins
-Terraform module to build Docker image with Jenkins, save it to an ECR repo, and deploy to Elastic Beanstalk running Docker stack
+# terraform-aws-jenkins
+
+Terraform module to build Docker image with [`Jenkins`](https://jenkins.io/), save it to an [`ECR`](https://aws.amazon.com/ecr/) repo, and deploy to [`Elastic Beanstalk`](https://aws.amazon.com/elasticbeanstalk/) running [`Docker`](https://www.docker.com/) stack
+
```
Comment (osterman) [inline README.md]: Add sample invocation.

File: main.tf [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+module "elastic_beanstalk_environment" {
+  source                  = "git::https://github.com/cloudposse/terraform-aws-elastic-beanstalk-environment.git?ref=tags/0.2.1"
+  attributes              = ["eb"]
+  namespace               = "${var.namespace}"
+  name                    = "${var.name}"
+  stage                   = "${var.stage}"
+  zone_id                 = "${var.zone_id}"
+  app                     = "${module.elastic_beanstalk_application.app_name}"
+  instance_type           = "${var.master_instance_type}"
+  autoscale_min           = 1
```
Comment (osterman) [inline main.tf]: Maybe add a comment as to why these are hardcoded and why they should not be changed.

Comment (aknysh) [reply main.tf]: Commented

File: README.md (hunk 5 above)
Comment (aknysh) [reply README.md]: Will create a complete `README.md` with examples

**Pull request merged**